src/irmin-pack/content_addressable_intf.ml

```diff
@@ -17,19 +17,27 @@
 open! Import
 
 module type S = sig
-  include Irmin.Content_addressable.S
+  type hash
 
-  val add : 'a t -> value -> key Lwt.t
+  include Irmin.Content_addressable.S with type hash := hash
+
+  val add : _ t -> value -> Key.t Lwt.t
   (** Overwrite [add] to work with a read-only database handler. *)
 
-  val unsafe_add : 'a t -> key -> value -> unit Lwt.t
+  val unsafe_add : _ t -> hash -> value -> Key.t Lwt.t
   (** Overwrite [unsafe_add] to work with a read-only database handler. *)
 
   val unsafe_append :
-    ensure_unique:bool -> overcommit:bool -> 'a t -> key -> value -> unit
+    ensure_unique_indexed:(* XXX: do we ever call this with [false]? *)
```

Comment:
for now, it is in the layered store https://github.com/mirage/irmin/blob/master/src/irmin-pack/layered/layered_store.ml#L45